src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -531,33 +539,103 @@ public void restoreState(List<UUID> list) throws Exception {
             // we can have more than one transaction to check in case of a scale-in event, or
             // for the reasons discussed in the 'notifyCheckpointComplete()' method.
 
-            if (list != null && list.size() > 0) {
-                for (UUID txnId : list) {
-                    if (txnId != null) {
-                        final Transaction<?> txn = pravegaWriter.getTxn(txnId);
-                        final Transaction.Status status = txn.checkStatus();
+            // we will create a writer per scope/stream from the pending transaction list and
+            // commit the transaction if it is still open
+
+            if (pendingTransactionList == null || pendingTransactionList.size() == 0) {
+                return;
+            }
+
+            Exception exception = null;
+
+            EventStreamWriter<T> restorePravegaWriter = null;
+
+            ClientFactory restoreClientFactory = null;
+
+            Map<String, Map<String, List<PendingTransaction>>> pendingTransactionsMap =
+                    pendingTransactionList.stream().collect(groupingBy(PendingTransaction::getScope, groupingBy(PendingTransaction::getStream)));
```

Comment:
Sorry to ask this, but could you rewrite it such that you group on the composite key of scope + stream?  That would eliminate the nested loop.  I think you can simply do:
```java
import io.pravega.client.stream.Stream;

Map<Stream, List<PendingTransaction>> pendingTransactionsMap = pendingTransactionList
  .stream()
  .collect(groupingBy(t -> Stream.of(t.getScope(), t.getStream());

```

Reply:
Yes, I have started with that approach but changed it since the `StreamImpl` class does not implement `equals()` and `hashcode()` methods which are necessary to uniquely identify the map key.

Reply:
It actually does thanks to the `@Data` annotation.  Personally I don't like Lombok for the reason that it is too subtle.